File: algorithms/cpp/_340_LongestSubstringWithAtMostKDistinctCharacters/_340_LongestSubstringWithAtMostKDistinctCharacters_std.cpp

# include <iostream>

using namespace std;


/**
* This reference program is provided by @jiuzhang.com
* Copyright is reserved. Please indicate the source for forwarding
*/

class Solution1 {
public:
    /**
     * @param s : A string
     * @return : The length of the longest substring
     *           that contains at most k distinct characters.
     */
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        // write your code here
        int start = 0, cnt = 0;
        int char_set[256] = {0};
        int ans = 0;
        for (int i = 0; i < s.size(); i++) {
            if (char_set[s[i]]++ == 0) cnt++;
            while (cnt > k) {
                char_set[s[start]]--;
                if (char_set[s[start++]] == 0) cnt--;
            }
            ans = max(i - start + 1, ans);
        }
        return ans;
    }
};

class Solution2
{
public:
    /**
     * @param s: A string
     * @param k: An integer
     * @return: An integer
     */
    int lengthOfLongestSubstringKDistinct(string &s, int k)
    {
        if(k <= 0)
            return 0;
        // 同向双指针，左指针为主指针
        int maxlen = 0, counter[128] = { 0 }, c = 0;
        for(const char *begin = s.c_str(), *end = begin; *end != '\0'; begin++)
        {
            while(*end != '\0' && (c < k || c == k && counter[*end] > 0))
            {
                if(counter[*end] == 0)
                    c++;
                counter[*end]++;
                end++;
            }
            maxlen = max(maxlen, (int)(end - begin));
            counter[*begin]--;
            if(counter[*begin] == 0)
                c--;
        }
        return maxlen;
    }
};
class Solution3 {
public:
    /**
     * @param s: A string
     * @param k: An integer
     * @return: An integer
     */
    int lengthOfLongestSubstringKDistinct(string &s, int k) {
        // write your code here
        int len = s.length();
        if (len <= k) return len;
        if (k == 0) return 0;
        //cout << (int)'z' << endl;

        int hash[256];
        memset(hash, 0, sizeof(int)*256);
        len = 1;
        int Max_len = 1;
        int num_char = 1;
        hash[s[0]] = 1;
        int start = 0;
        for (int i=1; i<s.length(); i++) {
            if (hash[s[i]] > 0) {
                hash[s[i]] ++;
                len ++;
                if (Max_len < len) Max_len = len;
            }
            else if (num_char < k){
                num_char ++;
                hash[s[i]] = 1;
                len ++;
                if (Max_len < len) Max_len = len;
            }
            else {
                while (num_char >= k) {
                    hash[s[start]] --;
                    if (hash[s[start]] == 0) num_char --;
                    len --;
                    start ++;
                }

                hash[s[i]] = 1;
                len ++;
                if (Max_len < len) Max_len = len;
                num_char ++;
            }
        }

        return Max_len;
    }
};
int main() {
    cout << "new file!" << endl;
    return 0;
}

File: algorithms/java/src/_340_LongestSubstringWithAtMostKDistinctCharacters/_340_LongestSubstringWithAtMostKDistinctCharacters_std.java
package _340_LongestSubstringWithAtMostKDistinctCharacters;

import java.util.HashMap;
import java.util.Map;

public class _340_LongestSubstringWithAtMostKDistinctCharacters_std {
    public static void main(String[] args) {
        _340_LongestSubstringWithAtMostKDistinctCharacters_std_ so = new _340_LongestSubstringWithAtMostKDistinctCharacters_std_();

    }
}

class _340_LongestSubstringWithAtMostKDistinctCharacters_std_ {

}

/**
 * This reference program is provided by @jiuzhang.com
 * Copyright is reserved. Please indicate the source for forwarding
 */

class Solution1 {
    /**
     * @param s : A string
     * @return : The length of the longest substring
     * that contains at most k distinct characters.
     */
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        // write your code here
        int maxLen = 0;

        // Key: letter; value: the number of occurrences.
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        int i, j = 0;
        char c;
        for (i = 0; i < s.length(); i++) {
            while (j < s.length()) {
                c = s.charAt(j);
                if (map.containsKey(c)) {
                    map.put(c, map.get(c) + 1);
                } else {
                    if (map.size() == k)
                        break;
                    map.put(c, 1);
                }
                j++;
            }

            maxLen = Math.max(maxLen, j - i);
            c = s.charAt(i);
            if (map.containsKey(c)) {
                int count = map.get(c);
                if (count > 1) {
                    map.put(c, count - 1);
                } else {
                    map.remove(c);
                }
            }
        }
        return maxLen;
    }
}


class Solution2 {
    /*
     * @param s: A string
     * @param k: An integer
     * @return: An integer
     */
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        // write your code here
        int[] cnt = new int[256];
        char[] sc = s.toCharArray();

        int ans = 0;
        int sum = 0;
        for (int l = 0, r = 0; r < s.length(); r++) {
            cnt[sc[r]]++;
            if (cnt[sc[r]] == 1) {
                sum++;
            }
            while (sum > k) {
                cnt[sc[l]]--;
                if (cnt[sc[l]] == 0) {
                    sum--;
                }
                l++;
            }
            ans = Math.max(ans, r - l + 1);
        }
        return ans;
    }
}
